File: search/연산자끼워넣기.py
# 연산자 끼워넣기 https://www.acmicpc.net/problem/14888

import sys

input = sys.stdin.readline
n = int(input())
nums = list(map(int, input().split()))
a, s, m, d = list(map(int, input().split()))

min_v = 2e9
max_v = -2e9


def dfs(cnt, now):
    global min_v, max_v, a, s, m, d
    if cnt == n:
        min_v = min(min_v, now)
        max_v = max(max_v, now)
        return
    if a > 0:
        a -= 1
        dfs(cnt + 1, now + nums[cnt])
        a += 1
    if s > 0:
        s -= 1
        dfs(cnt + 1, now - nums[cnt])
        s += 1
    if m > 0:
        m -= 1
        dfs(cnt + 1, now * nums[cnt])
        m += 1
    if d > 0:
        d -= 1
        dfs(cnt + 1, int(now / nums[cnt]))  # 음수 나누기 파이썬에서 자동 처리
        d += 1


dfs(1, nums[0])
print(max_v)
print(min_v)

File: implementation/연산자끼워넣기.js
// https://www.acmicpc.net/problem/14888

const fs = require("fs");
const inputs = fs.readFileSync("/dev/stdin").toString().split("\n");

const N = Number(inputs.shift());
let nums = inputs.shift().split(" ").map(Number);

let ops = new Map();
const op_chars = ["+", "-", "*", "/"];
const op_nums = inputs.shift().split(" ").map(Number);
for (let i = 0; i < op_chars.length; i++) {
  ops.set(op_chars[i], op_nums[i]);
}

// console.log(ops, nums)
let selected_ops = [];
let max_result = -Infinity;
let min_result = Infinity;
function dfs(cnt) {
  if (cnt === N - 1) {
    let result = nums[0];
    for (let i = 0; i < selected_ops.length; i++) {
      result = calc(result, selected_ops[i], nums[i + 1]);
    }
    max_result = Math.max(max_result, result);
    min_result = Math.min(min_result, result);
    return;
  }
  for (let i = 0; i < 4; i++) {
    let op = op_chars[i];
    if (ops.get(op) > 0) {
      ops.set(op, ops.get(op) - 1);
      selected_ops.push(op);
      dfs(cnt + 1);
      selected_ops.pop();
      ops.set(op, ops.get(op) + 1);
    }
  }
}

function calc(num1, op, num2) {
  switch (op) {
    case "+":
      return num1 + num2;
    case "-":
      return num1 - num2;
    case "*":
      return num1 * num2;
    case "/":
      if (num1 < 0) return parseInt(-num1 / num2) * -1;
      return parseInt(num1 / num2);
  }
}

dfs(0);

console.log(max_result);
console.log(min_result);
